apps/full-site-editing/full-site-editing-plugin/full-site-editing/templates/class-template-image-inserter.php

```diff
@@ -0,0 +1,171 @@
+<?php
+/**
+ * Full site editing file.
+ *
+ * @package A8C\FSE
+ */
+
+namespace A8C\FSE;
+
+/**
+ * Class Template_Image_Inserter
+ */
+class Template_Image_Inserter {
+	/**
+	 * Uploads specific images to the media libraray
+	 *
+	 * It then checks if any of the passed image URLs are also in the post content
+	 * of any of the passed post IDs. If so, then the post content is updated to
+	 * use the new local URL instead of the passed URL.
+	 *
+	 * @param array $img_urls URLs of images which should be uploaded.
+	 * @param array $post_ids An array of posts IDs which should be updated with newly uploaded image URLs.
+	 */
+	public function copy_images_and_update_posts( $img_urls = [], $post_ids = [] ) {
+		do_action(
+			'a8c_fse_log',
+			'template_image_sideload_begin',
+			[
+				'context'    => 'Template_Image_Inserter->copy_images_and_update_posts',
+				'image_urls' => $img_urls,
+			]
+		);
+
+		if ( empty( $img_urls ) || empty( $post_ids ) ) {
+			return;
+		}
+
+		$images = $this->upload_images( $img_urls );
+		foreach ( $post_ids as $id ) {
+			$this->update_post_images( $id, $images );
+		}
+	}
+
+	/**
+	 * Upload an array of images to the media library.
+	 *
+	 * If an file fails to upload, the error is logged to the FSE error handler.
+	 *
+	 * @param  array $image_urls Images URLs to upload.
+	 * @return array Each entry contains the old and new URL of an image if successfully uploaded.
+	 */
+	public function upload_images( $image_urls ) {
+		return array_reduce(
```

Comment:
This would be a bit on the "too clever" side, as I personally find reduces with long callbacks to be hard to read, but the inline comments make it very easy indeed! ✨

Reply:
I do agree it's too clever lol. I think I was trying to take your comment on the other PR into account and use more array-specific functions